docs/architecture/adr-012-ibc-relayer-v2.md

```diff
@@ -0,0 +1,670 @@
+without replacing existing features of the v1 relayer. In contrast, a v2
+relayer is expected to supercede the majority features of the v1 relayer
+with new and improved code.
+
+For the purpose of the architecture re-design, all the new code being
+developed are targeted toward the relayer v2. But the new code will be
+usable in the form of experimental features when the relayer v1.5 is
+released. Both the v1 relayer and the new relayer will co-exist from
+v1.5 onward, until the v2 relayer is released.
+
+In the relayer v1.5 MVP, the new relayer only re-implements the packet
+worker and the transaction sender. With that, the new relayer does not
+depend on the `RelayPath` and `Link` data types, as well as the `send_tx`
+methods in the v1 relayer. In contrast, the new relayer will initially
+continue to rely on the `ForeignClient` and `ChainHandle` datatypes
+to perform queries and processing of messages.
+
+## Architecture Overview
+
+A full description of the relayer v2 architecture is too much to be described
+in this ADR. Instead, readers are encouraged to read the full documentation
+from the generated Cargo docs for the new relayer crates.
+
+At its core, the v2 relayer makes use of a new programming technique, called
+_context-generic programming_ (CGP), to implement the relayer components in a
+modular fashion. CGP turns OOP methods into modular components by replacing
+the concrete `Self` type with a _generic_ `Context` type. Using the special
+properties of Rust's trait system, CGP allows component implementations to
+add new constraints and requirements to the `Context` type through `where`
+clauses inside `impl` blocks, by which the constraints would automatically
+propagated to the top level without having to repeatedly specify the
+constraints at every level of the code.
+
+### Relayer Framework
+
+Using CGP, the core logic of the new relayer is implemented as a fully
+abstract library crate called `ibc-relayer-framework`, with no dependency
+to external crates except for using the `async-trait` crate to support
+async functions inside traits. In addition, the relayer framework is
+developed with `#![no_std]` enabled. By having almost no external dependency,
+the relayer can be ported to various restrictive environments, such as
+Wasm, Substrate runtime, and symbolic execution environments.
+
+Since the relayer framework is fully abstract, it also does not depends on
+the concrete type definitions of the IBC constructs, including primitives
+like height. Instead, the types are declared as abstract associated types in
+traits like `HasChainTypes` and `HasRelayTypes`:
+
+```rust
+trait HasChainTypes {
+    type Height;
+    type Timestamp;
+    type Message;
+    type Event;
+    /* ... */
+}
+
+trait HasRelayTypes {
+    type SrcChain: HasChainTypes;
+    type DstChain: HasChainTypes;
+    type Packet;
+    /* ... */
+}
+```
+
+Readers are encouraged to refer to the documentation and source code for
```

Comment:
The tricky part for this is that the documentation has not been published to [docs.rs](https://docs.rs/), so there is no official link that we can refer to.

To workaround this, I aliased the ADR inside the cargo doc for `ibc-relayer-cosmos`, so that it can be rendered as a Rustdoc page together with the rest of the documentation. I also can't put it inside `ibc-relayer-framework`, because then the documentation wouldn't be able to link to `ibc-relayer-cosmos` or the relayer constructs in `ibc-relayer`.

Reply:
Here is the new rendered ADR as cargo doc: https://ibc-rs.netlify.app/ibc_relayer_cosmos/docs/adr/index.html